### PR TITLE
385 Financial aid state machine

### LINKED ISSUE #385: Financial aid state Machine Interface

### PR DESCRIPTION
**Pull request by dpoirier**
Depending on the current state of someone's application,
provide appropriate buttons on the dashboard for the user
to move their application to any other state that they
are allowed to move it to, given its current state.

E.g. if application is in submitted state, they can
withdraw it. If offered, they can accept or decline.
Etc.

Fixes #385

### CODE REVIEW

File: pycon/finaid/tests/test_views.py [hunk 1]
```diff
@@ -442,3 +443,162 @@ def test_two_applications(self):
...
+        self.assertEqual(200, rsp.status_code)
+        all_buttons = [
+            'finaid_apply', 'finaid_withdraw',
+            'finaid_decline', 'finaid_accept', 'finaid_request_more',
+            'finaid_edit', 'finaid_status', 'finaid_review',
+            'finaid_download_csv', 'finaid_provide_info'
+        ]
+        buttons_in_page = set()
+        for name in all_buttons:
+            text_repr, real_count, msg_prefix = self._assert_contains(
```
Comment (rebecca-caktus) [inline pycon/finaid/tests/test_views.py]: Interesting there. I hadn't seen `_assert_contains` used before.

Comment (dpoirier) [inline pycon/finaid/tests/test_views.py]: I had to dig into the implementation of assertContains to do what I wanted.

File: pycon/finaid/views.py [hunk 2]
```diff
@@ -359,6 +366,210 @@ def finaid_status(request):
 
 
 @login_required
+def finaid_accept(request):
```
Comment (rebecca-caktus) [inline pycon/finaid/views.py]: There seems to be a fair bit of repetition in these views. Would be worthwhile to distill a class-based base view?

File: pycon/templates/finaid/email/applicant/request_more/subject.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+{% load review_tags %}Message from {{ user.get_full_name|bleach|safe }} requesting more assistance.
```
Comment (rebecca-caktus) [inline pycon/templates/finaid/email/applicant/request_more/subject.txt]: ``review_tags`` don't seem to be used anywhere. (Ditto for the overridden subjects in the views file.)

Comment (rebecca-caktus) [inline pycon/templates/finaid/email/applicant/request_more/subject.txt]: Oh bleach.

**Comment by dpoirier** [comment]
@rebecca-caktus I've updated this to make the views DRYer.

File: pycon/finaid/views.py [hunk 4]
```diff
@@ -358,6 +367,159 @@ def finaid_status(request):
...
+        except FinancialAidReviewData.DoesNotExist:
+            FinancialAidReviewData(application=application)
+
+        application.set_status(self.new_status)
+        application.save()
+        message = FinancialAidMessage.objects.create(
+            user=request.user,
+            application=application,
+            visible=True,
+            message="The applicant has accepted their financial aid offer",
```
Comment (rebecca-caktus) [inline pycon/finaid/views.py]: Is this the same for all subclasses?

File: pycon/finaid/views.py [hunk 5]
```diff
@@ -358,6 +367,159 @@ def finaid_status(request):
...
+        return True
+
+
+class SendFinaidMessageViewBase(LoginRequiredMixin, View):
+    def dispatch(self, request, *args, **kwargs):
+        if not has_application(request.user):
+            messages.add_message(request, messages.ERROR,
+                                 _(u'You have not applied for financial aid'))
+            return redirect("dashboard")
+        if not self.status_okay(request):
```
Comment (rebecca-caktus) [inline pycon/finaid/views.py]: `status_okay` is not defined for this view.

File: pycon/finaid/views.py (hunk 4 above)
Comment (dpoirier) [inline pycon/finaid/views.py]: Fixed.

File: pycon/finaid/views.py (hunk 5 above)
Comment (dpoirier) [inline pycon/finaid/views.py]: Its subclasses all implement `status_okay`.  This view is not used directly.

**Comment by dpoirier** [comment]
Updated.

**Comment by dchukhin** [comment]
that's a lot of refactoring; looks good and DRYer

**Comment by hatlas** [comment]
Passed QA

**Pull request merged**